### PR TITLE
Implement string conversion for ReactionType

### PR DESCRIPTION
**Pull request by Marcuss2**

### CODE REVIEW

File: src/model/channel/reaction.rs [hunk 1]
```diff
@@ -3,23 +3,21 @@ use crate::http::CacheHttp;
...
-        Write as FmtWrite
-    },
-    str::FromStr
-};
+use std::{error::Error as StdError, fmt::{
+    Display,
+    Formatter,
+    Result as FmtResult,
+    Write as FmtWrite
+}, fmt};
```
Comment (arqunis) [inline src/model/channel/reaction.rs]: Unconventional formatting for imports.
```suggestion
use std::{
    error::Error as StdError, 
    fmt::{
        self,
        Display,
        Formatter,
        Result as FmtResult,
        Write as FmtWrite
    },
};
```

File: src/model/channel/reaction.rs [hunk 2]
```diff
@@ -397,11 +395,33 @@ impl From<EmojiIdentifier> for ReactionType {
...
+    type Error = ReactionConversionError;
+
+    fn try_from(emoji_string: String) -> std::result::Result<Self, Self::Error> {
+        if emoji_string.is_empty() {
+            return Err(ReactionConversionError)
+        }
+
+        if !emoji_string.starts_with('<') {
+            return Ok(ReactionType::Unicode(emoji_string))
+        }
```
Comment (arqunis) [inline src/model/channel/reaction.rs:423]: You're repeating the behaviour of the `TryFrom<&str>` impl redundantly.
```suggestion
```

File: src/model/channel/reaction.rs [hunk 3]
```diff
@@ -418,7 +438,68 @@ impl<'a> From<&'a str> for ReactionType {
...
+
+        if !emoji_str.starts_with('<') {
+            return Ok(ReactionType::Unicode(emoji_str.to_string()))
+        }
+
+        if !emoji_str.ends_with('>') {
+            return Err(ReactionConversionError);
+        }
+
```
Comment (arqunis) [inline src/model/channel/reaction.rs:482]:
```suggestion
let emoji_str = emoji_str.trim_matches(&['<', '>'] as &[char]);
```

File: src/model/channel/reaction.rs [hunk 4]
```diff
@@ -418,7 +438,68 @@ impl<'a> From<&'a str> for ReactionType {
...
+        let mut split_iter = emoji_str.split(':');
+
+        let animated = split_iter
+            .next()
+            .ok_or(ReactionConversionError)?
+            .chars()
+            .nth(1)
+            .unwrap_or('b') == 'a';
```
Comment (arqunis) [inline src/model/channel/reaction.rs]:
```suggestion
        let animated = split_iter.next().ok_or(ReactionConversionError)? == "a";
```

File: src/model/channel/reaction.rs [hunk 5]
```diff
@@ -418,7 +438,68 @@ impl<'a> From<&'a str> for ReactionType {
...
+            .to_string()
+            .into();
+
+        let id = split_iter
+            .next()
+            .ok_or(ReactionConversionError)?
+            .trim_end_matches('>')
+            .parse::<u64>()
+            .map_err(|_| ReactionConversionError)?
+            .into();
```
Comment (arqunis) [inline src/model/channel/reaction.rs]:
```suggestion
        let id = split_iter
            .next()
            .and_then(|s| s.parse::<u64>().ok())
            .ok_or(ReactionConversionError)?
            .into();
```

File: tests/test_reaction.rs [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+fn str_to_reaction_type_mangled_4() {
+    let emoji_str = "<:somestuff:1234";
+    ReactionType::try_from(emoji_str).unwrap_err();
+}
+
+#[test]
+fn str_fromstr() {
+    let emoji_str = "<:somestuff:1234";
+    ReactionType::from_str(emoji_str).unwrap_err();
+}
```
Comment (arqunis) [inline tests/test_reaction.rs]: Missing newline at the end of the file.

**Review by arqunis — CHANGES_REQUESTED** [review]

File: src/model/channel/reaction.rs (hunk 2 above)
Comment (Marcuss2) [reply src/model/channel/reaction.rs:423]: If the reaction is Unicode based, this avoids a string allocation

File: src/model/channel/reaction.rs [hunk 7]
```diff
@@ -418,7 +442,61 @@ impl<'a> From<&'a str> for ReactionType {
...
+        }
+
+        if !emoji_str.starts_with('<') {
+            return Ok(ReactionType::Unicode(emoji_str.to_string()))
+        }
+
+        if !emoji_str.ends_with('>') {
+            return Err(ReactionConversionError);
+        }
+let emoji_str = emoji_str.trim_matches(&['<', '>'] as &[char]);
```
Comment (arqunis) [inline src/model/channel/reaction.rs]:
```suggestion

    let emoji_str = emoji_str.trim_matches(&['<', '>'] as &[char]);

```

File: src/model/channel/reaction.rs [hunk 8]
```diff
@@ -418,7 +442,61 @@ impl<'a> From<&'a str> for ReactionType {
...
-    fn from(unicode: &str) -> ReactionType { ReactionType::Unicode(unicode.to_string()) }
+    ///
+    /// Creating a `ReactionType` from a custom emoji argument in the following format:
+    ///
+    /// ```rust
+    /// use serenity::model::channel::ReactionType;
+    /// use serenity::model::id::EmojiId;
+    ///
+    /// let emoji_string = "<:customemoji:600404340292059257>";
+    /// let reaction = ReactionType::from(emoji_string);
```
Comment (arqunis) [inline src/model/channel/reaction.rs]:
```suggestion
    /// let reaction = ReactionType::try_from(emoji_string).unwrap();
```

File: src/model/channel/reaction.rs [hunk 9]
```diff
@@ -397,11 +399,33 @@ impl From<EmojiIdentifier> for ReactionType {
...
 }
 
-impl From<String> for ReactionType {
-    fn from(unicode: String) -> ReactionType { ReactionType::Unicode(unicode) }
+#[derive(Debug)]
+pub struct ReactionConversionError;
+
+impl Display for ReactionConversionError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "Failed to convert String to Reaction")
```
Comment (arqunis) [inline src/model/channel/reaction.rs]:
```suggestion
        write!(f, "failed to convert from a string to ReactionType")
```

**Review by arqunis — CHANGES_REQUESTED** [review]

**Review by arqunis — APPROVED** [review]

**Comment by arqunis** [comment]
For the purposes of fixing one last formatting issue and squashing the commits, I have taken the liberty of merging this PR manually.

https://github.com/serenity-rs/serenity/commit/6aee47b93d7fdead6eb9e4b3cc7269fd0fc45619

**Pull request closed without merging**